### PR TITLE
Make a function 'constexpr'.

### PR DESCRIPTION
**Pull request by bangerth**
Marking the function as `constexpr` means that it also needs to be defined in the `.h` file.

I will need this for a later patch where I make the special `ReferenceCell`
objects `constexpr` as well. This patch is simply meant to make that later
review marginally easier.

/rebuild

### CODE REVIEW

**Comment by bangerth** [comment]
See #11544.

**Review by peterrum — APPROVED** [review]

File: include/deal.II/grid/reference_cell.h [hunk 1]
```diff
@@ -338,6 +338,24 @@ ReferenceCell::operator!=(const ReferenceCell &type) const
 
 
 
+namespace internal
+{
+  namespace ReferenceCell
+  {
+    constexpr dealii::ReferenceCell
```
Comment (masterleinad) [inline include/deal.II/grid/reference_cell.h:345]:
```suggestion
    DEAL_II_CONSTEXPR dealii::ReferenceCell
```

File: include/deal.II/grid/reference_cell.h [hunk 2]
```diff
@@ -60,7 +60,7 @@ namespace internal
      * but we have this one function in an internal namespace that is a friend
      * of the class and can be used to create the objects.
      */
-    dealii::ReferenceCell
+    constexpr dealii::ReferenceCell
```
Comment (masterleinad) [inline include/deal.II/grid/reference_cell.h:63]:
```suggestion
    DEAL_II_CONSTEXPR dealii::ReferenceCell
```

File: include/deal.II/grid/reference_cell.h [hunk 3]
```diff
@@ -297,7 +297,7 @@ class ReferenceCell
    * A kind of constructor -- not quite private because it can be
    * called by anyone, but at least hidden in an internal namespace.
    */
-  friend ReferenceCell
+  friend constexpr ReferenceCell
```
Comment (masterleinad) [inline include/deal.II/grid/reference_cell.h:300]:
```suggestion
  friend DEAL_II_CONSTEXPR ReferenceCell
```

**Review by masterleinad — COMMENTED** [review]
We call Assert in these functions and some compilers don't like that so we explicitly check for that and define DEAL_II_CONSTEXPR accordingly.

**Comment by bangerth** [comment]
Ah, yes, that's a bummer. Let's try this again...

**Review by masterleinad — APPROVED** [review]

File: include/deal.II/grid/reference_cell.h [hunk 4]
```diff
@@ -338,6 +338,27 @@ ReferenceCell::operator!=(const ReferenceCell &type) const
 
 
 
+namespace internal
+{
+  namespace ReferenceCell
+  {
+    DEAL_II_CONSTEXPR dealii::ReferenceCell
```
Comment (masterleinad) [inline include/deal.II/grid/reference_cell.h]:
```suggestion
    inline DEAL_II_CONSTEXPR dealii::ReferenceCell
```

Comment (masterleinad) [reply include/deal.II/grid/reference_cell.h]: Wow. You pushed the suggestion the same time I clicked "Comment". :smile:

Comment (bangerth) [reply include/deal.II/grid/reference_cell.h]: Yup, discovered that a couple of minutes ago as well. Let's see whether this iteration works -- I'm glad I broke this patch out of the bigger one, it turns out to be unexpectedly difficult...

Comment (bangerth) [reply include/deal.II/grid/reference_cell.h]: We're on it in this project :-)

**Pull request merged**